nucypher/network/retrieval.py

```diff
@@ -0,0 +1,410 @@
+
+
+class ReencryptionResponse:
+    """
+    A response from Ursula with reencrypted capsule frags.
+    """
+
+    @classmethod
+    def construct_by_ursula(cls,
+                            capsules: List[Capsule],
+                            cfrags: List[VerifiedCapsuleFrag],
+                            stamp: SignatureStamp,
+                            ) -> 'ReencryptionResponse':
+
+        # un-verify
+        cfrags = [CapsuleFrag.from_bytes(bytes(cfrag)) for cfrag in cfrags]
+
+        capsules_bytes = b''.join(bytes(capsule) for capsule in capsules)
+        cfrags_bytes = b''.join(bytes(cfrag) for cfrag in cfrags)
+        signature = stamp(capsules_bytes + cfrags_bytes)
+        return cls(cfrags, signature)
+
+    def __init__(self, cfrags: List[CapsuleFrag], signature: Signature):
+        self.cfrags = cfrags
+        self.signature = signature
+
+    @classmethod
+    def from_bytes(cls, data: bytes):
+        signature, cfrags_bytes = signature_splitter(data, return_remainder=True)
+
+        # We would never send a request with no capsules, so there should be cfrags.
+        # The splitter would fail anyway, this just makes the error message more clear.
+        if not cfrags_bytes:
+            raise ValueError("ReencryptionResponse contains no cfrags")
+
+        cfrags = cfrag_splitter.repeat(cfrags_bytes)
+        return cls(cfrags, signature)
+
+    def __bytes__(self):
+        return bytes(self.signature) + b''.join(bytes(cfrag) for cfrag in self.cfrags)
+
+
+class RetrievalClient:
+    """
+    Capsule frag retrieval machinery shared between Bob and Porter.
+    """
+
+    def __init__(self, learner: Learner):
+        self._learner = learner
+        self.log = Logger(self.__class__.__name__)
+
+    def _ensure_ursula_availability(self, treasure_map: TreasureMap, timeout=10):
+        """
+        Make sure we know enough nodes from the treasure map to decrypt;
+        otherwise block and wait for them to come online.
+        """
+
+        # OK, so we're going to need to do some network activity for this retrieval.
+        # Let's make sure we've seeded.
+        if not self._learner.done_seeding:
+            self._learner.learn_from_teacher_node()
+
+        ursulas_in_map = treasure_map.destinations.keys()
+        all_known_ursulas = self._learner.known_nodes.addresses()
+
+        # Push all unknown Ursulas from the map in the queue for learning
+        unknown_ursulas = ursulas_in_map - all_known_ursulas
+        if unknown_ursulas:
+            self._learner.learn_about_specific_nodes(unknown_ursulas)
+
+        # How many nodes over the threshold we want to know (just in case)
+        redundancy = 0
+
+        # If we know enough to decrypt, we can proceed.
+        known_ursulas = ursulas_in_map & all_known_ursulas
+
+        if len(known_ursulas) - redundancy >= treasure_map.threshold:
+            return
+
+        allow_missing = max(len(unknown_ursulas) - redundancy, 0)
```

Comment:
This calculation seems off or am I missing something? If you get here the `len(known_ursulas)` is not sufficient to make the threshold (L279).

Now, suppose shares are 8, and threshold is 7. Also suppose that `len(known_ursulas)` is 5 i.e. not met the threshold. Therefore, `len(unknown_ursulas) = shares - len(known_ursulas) = 8 - 5 = 3`

Therefore `allow_missing` gets set to `max(3 - 0, 0) = 3`, whereas technically it should be 1 because out of the 3 unknown we need 2 to meet the threshold, and it's ok if we don't learn about 1 i.e. `len(unknown_ursulas) - (threshold - len(known_ursulas))` = 3 - (7 - 5) = 1

Reply:
You are right, my bad. Adjusted the formula.